### PR TITLE
[core] Minor optimization on message size

### PR DESCRIPTION
**Pull request by dentiny**
I found we repeatedly calculate protobuf message size during message publish, which requires recursive size calculation on message fields and I found it bottleneck in the past.

### CODE REVIEW

File: src/ray/pubsub/publisher.cc [hunk 1]
```diff
@@ -22,19 +22,17 @@ namespace pubsub {
 
 namespace pub_internal {
 
-bool EntityState::Publish(std::shared_ptr<rpc::PubMessage> msg) {
+bool EntityState::Publish(std::shared_ptr<rpc::PubMessage> msg, uint64_t msg_size) {
```
Comment (jjyao) [inline src/ray/pubsub/publisher.cc:25]: size_t

```

virtual size_t | ByteSizeLong() constComputes the serialized size of the message. more...
-- | --


virtual size_t	
ByteSizeLong() const
Computes the serialized size of the message. [more...](https://protobuf.dev/reference/cpp/api-docs/google.protobuf.message/#Message.ByteSizeLong.details)
```

Comment (dentiny) [reply src/ray/pubsub/publisher.cc:25]: I used `uint64_t` since it's used elsewhere...

File: src/ray/pubsub/publisher.h [hunk 2]
```diff
@@ -73,17 +73,17 @@ class EntityState {
...
   // Protobuf messages fail to serialize if 2GB or larger. Cap published
   // message batches to this size to ensure that we can publish each message
   // batch. Individual messages larger than this limit will also be dropped.
   // TODO(swang): Pubsub clients should also ensure that they don't try to
   // publish messages larger than this.
-  const int64_t max_message_size_bytes_;
+  const size_t max_message_size_bytes_;
   // Set to -1 to disable buffering.
-  const int64_t max_buffered_bytes_;
+  const size_t max_buffered_bytes_;
```
Comment (jjyao) [inline src/ray/pubsub/publisher.h:84]: Based on the comment we cannot use size_t since there is `-1`?

File: src/ray/pubsub/publisher.h [hunk 3]
```diff
@@ -73,17 +74,17 @@ class EntityState {
...
   // Size of each inflight message.
-  std::queue<int64_t> message_sizes_;
+  std::queue<size_t> message_sizes_;
   // Protobuf messages fail to serialize if 2GB or larger. Cap published
   // message batches to this size to ensure that we can publish each message
   // batch. Individual messages larger than this limit will also be dropped.
   // TODO(swang): Pubsub clients should also ensure that they don't try to
   // publish messages larger than this.
-  const int64_t max_message_size_bytes_;
+  const size_t max_message_size_bytes_;
```
Comment (jjyao) [inline src/ray/pubsub/publisher.h:83]: Can  you change `RAY_CONFIG(int64_t, max_grpc_message_size, 512 * 1024 * 1024)` to `size_t` as well?

**Review by jjyao — APPROVED** [review]

File: src/ray/pubsub/publisher.h (hunk 3 above)
Comment (dentiny) [reply src/ray/pubsub/publisher.h:83]: done

**Comment by jjyao** [comment]
Build failures.

**Comment by dentiny** [comment]
> Build failures.

Interesting, it actually builds without warning on my local
```

[2024-12-30T23:14:45Z] src/ray/pubsub/publisher.cc:336:67: error: comparison of integer expressions of different signedness: 'int64_t' {aka 'long int'} and 'size_t' {aka 'long unsigned int'} [-Werror=sign-compare]
--
  | [2024-12-30T23:14:45Z]   336 \|       if (num_total_bytes > 0 && num_total_bytes + msg_size_bytes >
  | [2024-12-30T23:14:45Z]       \|                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  | [2024-12-30T23:14:45Z]   337 \|                                      RayConfig::instance().max_grpc_message_size()) {
  | [2024-12-30T23:14:45Z]       \|                                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  | [2024-12-30T23:14:45Z] cc1plus: all warnings being treated as errors


```

**Comment by jjyao** [comment]
Tests failed to build.

**Comment by dentiny** [comment]
Hi @jjyao could you please help me merge it? Thanks!

**Pull request merged**